iroha_introspect/src/lib.rs

```diff
@@ -0,0 +1,303 @@
+//! Module for introspecting rust types in other languages for translation.
+
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    time::Duration,
+};
+
+/// Derive introspect. It will make your structure introspectable
+pub use iroha_introspect_derive::Introspect;
+use serde::Serialize;
+
+/// Metadata map
+pub type MetaMap = BTreeMap<String, Metadata>;
+
+/// Introspect trait
+pub trait Introspect {
+    /// Returns unique type name.
+    /// WARN: `std::any::type_name` is compiler related, so is not unique.
+    /// I guess we should change it somehow later
+    fn type_name() -> String {
+        let mut name = module_path!().to_owned();
+        name.push_str("::");
+        name.push_str(std::any::type_name::<Self>());
+        name
+    }
+
+    /// Returns info about current type. Will return map from type names to its metadata
+    fn get_introspect() -> MetaMap {
+        let mut map = MetaMap::new();
+        Self::introspect(&mut map);
+        map
+    }
+
+    /// Introspect function. Give it empty map and it will return description of types
+    /// related to this type
+    fn introspect(metamap: &mut MetaMap);
+}
+
+/// Metadata
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub enum Metadata {
+    /// Named structure
+    NamedStruct(NamedStructMeta),
+    /// Unnamed structure
+    UnnamedStruct(UnnamedStructMeta),
+    /// Enum
+    Enum(EnumMeta),
+    /// Integer
+    Int(IntMode),
+    /// String
+    String,
+    /// Bool
+    Bool,
+    /// Vector with type
+    Vec(String),
+    /// Option with type
+    Option(String),
+    /// Map
+    Map(MapMeta),
+    /// Result
+    Result(ResultMeta),
+}
+
+/// Named structure. It has names on fields
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct NamedStructMeta {
+    /// Fields
+    pub declarations: Vec<Declaration>,
+    //todo add collection of properties meta defined in struct
+}
+
+/// Field
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct Declaration {
+    /// Field name
+    pub name: String,
+    /// Type
+    pub ty: String,
+}
+
+/// Unnamed structure. It doesn't have name of fields
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct UnnamedStructMeta {
```

Comment:
This is ambiguous. Do we really need to have such variant?Previous version with Option<String> with None for tuple-like structs doesn't acceptable here? Btw if struct do not have any fields (unit struct) it will be NamedStruct or UnnamedStruct?

Reply:
I mean it better encodes rust type system. If 1 field doesn't have name then others also can't have it.

Reply:
Zero size structure and tuples will be unnamed structure

Reply:
I think we need to hide rust type-system in scheme but not expose it, because it is not friendly for  language without such features. I think we need to try to make our scheme language-agnostic as much as possible.

Reply:
It won't be language agnostic in any way. It is like encode json in language agnostic way. It won't make sense.

Using previous variant would allow invalid rust types, so we could remove invalid states by this variant

Reply:
"UnnamedStruct" is definitely bad name. At least lets rename it

Reply:
May be NamedFields and UnnamedFields?

Reply:
Yes, that is better